File: app/api/src/main/kotlin/com/wallet/clover/api/service/ExtractionService.kt
package com.wallet.clover.api.service

import com.wallet.clover.api.domain.extraction.ExtractionContext
import com.wallet.clover.api.dto.ExtractionRequest
import com.wallet.clover.api.service.LottoNumberExtractor
import org.springframework.stereotype.Service

/**
 * 로또 번호 추출 기능을 제공하는 애플리케이션 서비스입니다.
 * 다양한 미신적 방법론에 따라 로또 번호를 생성합니다.
 */
@Service
class ExtractionService(
    private val lottoNumberExtractor: LottoNumberExtractor,
) {
    /**
     * 지정된 추출 방법에 따라 로또 번호를 생성합니다.
     *
     * @param request 추출 요청 정보 (방법 및 컨텍스트 데이터)
     * @return 생성된 로또 번호 6개 Set
     */
    suspend fun extractLottoNumbers(request: ExtractionRequest): Set<Int> {
        val context = ExtractionContext(
            dreamKeyword = request.dreamKeyword,
            birthDate = request.birthDate,
            personalKeywords = request.personalKeywords,
            natureKeyword = request.natureKeyword,
            divinationKeyword = request.divinationKeyword,
            colorKeyword = request.colorKeyword,
            animalKeyword = request.animalKeyword,
        )
        return lottoNumberExtractor.extract(request.method, context)
    }
}

File: backend-node/src/lotto/extraction.service.ts
import { Injectable } from '@nestjs/common';
import {
  LottoNumberExtractor,
  ExtractionContext,
} from './lotto-number.extractor';
import { ExtractNumbersDto } from './dto/extract-numbers.dto';

/**
 * 다양한 방식(꿈 해몽, 사주 등)으로 로또 번호를 생성하는 요청을 처리하는 서비스입니다.
 */
@Injectable()
export class ExtractionService {
  /**
   * 로또 번호 추출기 인스턴스를 주입받습니다.
   * @param extractor 실제 번호 생성 로직을 담당하는 컴포넌트
   */
  constructor(private readonly extractor: LottoNumberExtractor) {}

  /**
   * 요청된 방식에 따라 로또 번호 6개를 생성합니다.
   * @param dto 추출 방식 및 관련 키워드 정보
   */
  async extractLottoNumbers(dto: ExtractNumbersDto): Promise<number[]> {
    const context: ExtractionContext = {
      dreamKeyword: dto.dreamKeyword,
      birthDate: dto.birthDate,
      personalKeywords: dto.personalKeywords,
      natureKeyword: dto.natureKeyword,
      divinationKeyword: dto.divinationKeyword,
      colorKeyword: dto.colorKeyword,
      animalKeyword: dto.animalKeyword,
    };
    return this.extractor.extract(dto.method, context);
  }
}
